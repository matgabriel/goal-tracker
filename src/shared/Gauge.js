// Jauge
// =====
//
// Réutilisé absolument partout dans l’application…  Un parfait exemple de
// composant réutilisable, du coup !

import { hot } from 'react-hot-loader/root'
import React from 'react'

import LinearProgress from '@material-ui/core/LinearProgress'

import { nonNegativeInteger, positiveInteger } from '../shared/prop-types'

const Gauge = ({ max, value }) => {
  return (
    <LinearProgress
      style={{ height: 8 }}
      variant='determinate'
      value={normalize(value, max)}
    />
  )
}

// En définissant les valeurs par défaut dans `defaultProps` au lieu des valeurs
// par défaut de la signature, on s’assure qu’elles seront bien prises en compte
// avant l’exploitation des `propTypes` ci-après.
Gauge.defaultProps = {
  max: 100,
}

// Comme toujours, on définit les propriétés attendues/autorisées pour
// validation.
Gauge.propTypes = {
  max: positiveInteger,
  value: nonNegativeInteger.isRequired,
}

function normalize(value, max) {
  return value * (100 / max)
}

export default hot(Gauge)
