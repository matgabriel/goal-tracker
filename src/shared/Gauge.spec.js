import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Gauge from './Gauge'

describe('<Gauge />', () => {
  it('should render appropriately', () => {
    const { getByRole } = render(<Gauge value={50} />)
    const progressBar = getByRole('progressbar')
    expect(progressBar).toHaveClass('MuiLinearProgress-determinate')
    expect(progressBar).toHaveAttribute(
      'style',
      expect.stringContaining('height: 8px')
    )
    expect(progressBar).toHaveAttribute('aria-valuenow', '50')
  })

  it('should normalize value on custom max', () => {
    const { getByRole } = render(<Gauge value={40} max={80} />)
    expect(getByRole('progressbar')).toHaveAttribute('aria-valuenow', '50')
  })

  it('should otherwise match the expected snapshot', () => {
    const { container } = render(<Gauge value={50} />)

    expect(container).toMatchSnapshot()
  })
})
