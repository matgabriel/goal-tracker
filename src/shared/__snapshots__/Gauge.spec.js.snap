// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gauge /> should otherwise match the expected snapshot 1`] = `
<div>
  <div
    aria-valuenow="50"
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
    role="progressbar"
    style="height: 8px;"
  >
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;
